services/slam/builtin/cartographer_int_test.go

```diff
@@ -156,8 +176,136 @@ func TestCartographerIntegration(t *testing.T) {
 
 	// Close out slam service
 	test.That(t, utils.TryClose(context.Background(), svc), test.ShouldBeNil)
+
+	// Don't clear out the directory, since we will re-use the maps for the next run
+	closeOutSLAMService(t, "")
+
+	// added sleep to ensure cartographer stops
+	time.Sleep(time.Millisecond * cartoSleepMS)
```

Comment:
So... if you're actually calling Close(), that close is returning no error, and yet it's still "doing stuff" afterwards, I think that points to a bug, one you should probably hunt down, rather than work around it here.

My guess is that, since there's a lot of background go routines, there's probably an ordering problem between the workgroup sync stuff (activeBackgroundWorkers.) One specific thing that _might_ be it is the ticker in StartDataProcess, which is deferred before the wg.Done() call is deferred, meaning the ticker may still fire AFTER the workgroup is marked done (in a race condition.) I could be wrong on that, but it's possible.

Reply:
@Otterverse, I don't think that would be the issue here, since the thing that is "doing stuff" after close is the `carto_grpc_server` binary. This is managed as a `ManagedProcess` and gets shut down by `StopSLAMProcess()`. I think it would be worth taking an hour to investigate what is happening during https://github.com/viamrobotics/goutils/blob/2d1d4397829eed318a3bfb7e2bc8e0427a71c229/pexec/managed_process.go#L287, but if we can't figure it out, I'm okay with this as a workaround. What do you think?

Reply:
I've been working a lot with Managed Process, and it does have some issues stopping (grand) child processes correctly. (I'm working on a PR for that.) That _could_ be a source of it. So if the case that's happening here is for sure that the binary process is what's still creating stuff after Close(), then maybe it is that (or something in the appimage layer(s), ugh.)

That said, if your team is comfortable with a 100ms delay as a workaround, (and the workaround is reliable) I won't block on this. I would like to see it documented as an issue for the future in that case though.

Reply:
yeah looking through the test logs linked above, cartographer never actually prints its "[finishing session](https://github.com/viamrobotics/slam/blob/main/slam-libraries/viam-cartographer/src/main.cc#L14)" log that we expect to see when exiting the server. So I am under the assumption that the Close() call isn't actually working, which is why we get the "[unable to check exit status](https://github.com/viamrobotics/rdk/actions/runs/3687572584/jobs/6241321866#step:6:474)" log. Also worth noting that the timer for the stopping and force exit log does not seem to last 3 seconds like its [configured](https://github.com/viamrobotics/goutils/blob/2d1d4397829eed318a3bfb7e2bc8e0427a71c229/pexec/managed_process.go#L23) to be

Reply:
Do you know why we're not getting an error back from Close()?

Reply:
"unable to check exit status" shows up IF the process got killed instead of closing on its own sometimes. We currently have a hardcoded wait of only 3 seconds between sending SIGINT and then SIGKILL (which leaves no exit code.) However, the kill should happen before the process.Stop() returns. But as noted, it doesn't kill children of children, so it could be orphaning processes.

I'm going to put this in my todo to try to reproduce and look into later this week if I have time.

Reply:
In the meantime, if you want to try using my "fix" branch in your local testing, you can add this to go.mod

`replace go.viam.com/utils => github.com/Otterverse/goutils b7bf775`

The PR for it is here (not quite finished yet, needs more tests.) https://github.com/viamrobotics/goutils/pull/116

Reply:
the go.mod file is telling me I need to have the version in the form of `v1.2.3` instead of  `b7bf775`, what should I put for that?

Reply:
Doh! Forgot when you use that short syntax, you have to run `go mod tidy` and then it'll download and correct the syntax. (I'm just in the habit of always running that after I modify go.mod)